### PR TITLE
Improve identity projection when the selectedPositions is a list

### PR DESCRIPTION
**Pull request by yingsu00**
This PR optimizes identity projection with two ways:

1. Create DictionaryBlock when the selectedPositions is a list;
2. Use a faster and more memory efficient block size calculation.

BenchmarkPageProcessor shows 21x improvement in throughput, 46x reduction in gc.alloc.rate.norm, and 9x reduction in
GC count and time:

Before

```
    Benchmark                                                                    Mode  Cnt       Score      Error   Units
    BenchmarkPageProcessor.identityProjection                                   thrpt   50  125261.816 ± 5342.200   ops/s
    BenchmarkPageProcessor.identityProjection:·gc.alloc.rate                    thrpt   50    2439.004 ±  103.990  MB/sec
    BenchmarkPageProcessor.identityProjection:·gc.alloc.rate.norm               thrpt   50   40832.029 ±    0.033    B/op
    BenchmarkPageProcessor.identityProjection:·gc.churn.PS_Eden_Space           thrpt   50    2461.428 ±  364.315  MB/sec
    BenchmarkPageProcessor.identityProjection:·gc.churn.PS_Eden_Space.norm      thrpt   50   41439.660 ± 6521.978    B/op
    BenchmarkPageProcessor.identityProjection:·gc.churn.PS_Survivor_Space       thrpt   50       0.016 ±    0.016  MB/sec
    BenchmarkPageProcessor.identityProjection:·gc.churn.PS_Survivor_Space.norm  thrpt   50       0.257 ±    0.267    B/op
    BenchmarkPageProcessor.identityProjection:·gc.count                         thrpt   50      46.000             counts
    BenchmarkPageProcessor.identityProjection:·gc.time                          thrpt   50     181.000                 ms
```

After

```
    Benchmark                                                                    Mode  Cnt        Score        Error   Units
    BenchmarkPageProcessor.identityProjection                                   thrpt   10  2687127.250 ± 337956.918   ops/s
    BenchmarkPageProcessor.identityProjection:·gc.alloc.rate                    thrpt   10     1118.285 ±    141.900  MB/sec
    BenchmarkPageProcessor.identityProjection:·gc.alloc.rate.norm               thrpt   10      872.001 ±      0.005    B/op
    BenchmarkPageProcessor.identityProjection:·gc.churn.PS_Eden_Space           thrpt   10     1050.013 ±   1683.604  MB/sec
    BenchmarkPageProcessor.identityProjection:·gc.churn.PS_Eden_Space.norm      thrpt   10      827.947 ±   1329.288    B/op
    BenchmarkPageProcessor.identityProjection:·gc.churn.PS_Survivor_Space       thrpt   10        0.012 ±      0.040  MB/sec
    BenchmarkPageProcessor.identityProjection:·gc.churn.PS_Survivor_Space.norm  thrpt   10        0.011 ±      0.034    B/op
    BenchmarkPageProcessor.identityProjection:·gc.count                         thrpt   10        5.000               counts
    BenchmarkPageProcessor.identityProjection:·gc.time                          thrpt   10       21.000                   ms
```

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

File: presto-common/src/main/java/com/facebook/presto/common/block/AbstractMapBlock.java [hunk 1]
```diff
@@ -194,6 +194,22 @@ public long getRegionLogicalSizeInBytes(int position, int length)
...
+        int positionCount = getPositionCount();
+        checkValidRegion(positionCount, position, length);
+
+        int entriesStart = getOffset(position);
+        int entriesEnd = getOffset(position + length);
+        int entryCount = entriesEnd - entriesStart;
+
+        return getRawKeyBlock().getApproximateRegionLogicalSizeInBytes(entriesStart, entryCount) +
+                getRawValueBlock().getApproximateRegionLogicalSizeInBytes(entriesStart, entryCount) +
+                (Integer.BYTES + Byte.BYTES) * positionCount +
```
Comment (mbasmanova) [inline presto-common/src/main/java/com/facebook/presto/common/block/AbstractMapBlock.java:209]: positionCount -> entryCount or length?

File: presto-common/src/main/java/com/facebook/presto/common/block/AbstractRowBlock.java [hunk 2]
```diff
@@ -130,6 +129,24 @@ public long getRegionLogicalSizeInBytes(int position, int length)
...
+    public long getApproximateRegionLogicalSizeInBytes(int position, int length)
+    {
+        int positionCount = getPositionCount();
+        checkValidRegion(positionCount, position, length);
+
+        int startFieldBlockOffset = getFieldBlockOffset(position);
+        int endFieldBlockOffset = getFieldBlockOffset(position + length);
+        int fieldBlockLength = endFieldBlockOffset - startFieldBlockOffset;
+
+        long approximateLogicalSizeInBytes = (Integer.BYTES + Byte.BYTES) * positionCount;
```
Comment (mbasmanova) [inline presto-common/src/main/java/com/facebook/presto/common/block/AbstractRowBlock.java:142]: positionCount -> length?

File: presto-common/src/main/java/com/facebook/presto/common/block/DictionaryBlock.java [hunk 3]
```diff
@@ -280,6 +280,12 @@ public long getRegionLogicalSizeInBytes(int positionOffset, int length)
         return sizeInBytes;
     }
 
+    @Override
+    public long getApproximateRegionLogicalSizeInBytes(int position, int length)
+    {
+        return (long) (dictionary.getApproximateRegionLogicalSizeInBytes(0, dictionary.getPositionCount()) * 1.0f * length / positionCount);
```
Comment (mbasmanova) [inline presto-common/src/main/java/com/facebook/presto/common/block/DictionaryBlock.java:286]: what's the reason for `* 1.0f` ?

**Review by mbasmanova — COMMENTED** [review]
@yingsu00 Some questions below.

File: presto-common/src/main/java/com/facebook/presto/common/block/AbstractMapBlock.java (hunk 1 above)
Comment (yingsu00) [reply presto-common/src/main/java/com/facebook/presto/common/block/AbstractMapBlock.java:209]: @mbasmanova thanks for reading so carefully! It shoudl be length. I also added the comments
```
                (Integer.BYTES + Byte.BYTES) * length +         // offsets and mapIsNull
                Integer.BYTES * HASH_MULTIPLIER * entryCount;   // hashtables
```

File: presto-common/src/main/java/com/facebook/presto/common/block/DictionaryBlock.java (hunk 3 above)
Comment (yingsu00) [reply presto-common/src/main/java/com/facebook/presto/common/block/DictionaryBlock.java:286]: You are right it doesn't matter. I removed it and it's just
```
return dictionary.getApproximateRegionLogicalSizeInBytes(0, dictionary.getPositionCount()) * length / positionCount;
```

**Comment by yingsu00** [comment]
@mbasmanova just FYI I'm fixing the assertBlockSize() test to cover the case mistakenly used `positionCount` instead of `length` in Map and RowBlocks. Will leave a message once it's done.

File: presto-common/src/main/java/com/facebook/presto/common/block/AbstractMapBlock.java (hunk 1 above)
Comment (mbasmanova) [reply presto-common/src/main/java/com/facebook/presto/common/block/AbstractMapBlock.java:209]: @yingsu00 Thanks for adding comments. These make it very easy to understand what the code is doing and why.

**Comment by yingsu00** [comment]
@mbasmanova Masha, I have updated commit "Add getApproximateRegionLogicalSizeInBytes() to Block " and it's ready for review again. Thank you very much for your careful reading!

**Review by mbasmanova — APPROVED** [review]

**Pull request merged**